src/logger/logger.ts

```diff
@@ -1,12 +1,139 @@
-import pino from "pino"
+import _ from "lodash"
+import pino, { Logger as PinoLogger } from "pino"
+
+import { StackFrame, get, parse } from "./stack-trace"
+
+interface WarnMeta {
+  params: Record<string, unknown>
+}
+
+export const MAX_STACK_DEPTH = 5
+
+export interface TraceMeta {
+  file: string
+  line: number
+  func: string
+}
+
+export const getTraceMeta = (frames: StackFrame[]): TraceMeta[] =>
+  frames.map((frame) => ({
+    file: frame.getFileName(),
+    line: frame.getLineNumber(),
+    func: frame.getFunctionName(),
+  }))
+
+const hasErrorProp = (
+  messageOrError: unknown
+): messageOrError is { error: Error } =>
+  typeof messageOrError === "object" &&
+  !!messageOrError &&
+  "error" in messageOrError
+
+export class Logger {
+  _logger: PinoLogger
+
+  constructor(logger: PinoLogger) {
+    this._logger = logger
+  }
+
+  public child = (
+    bindings: Record<string, unknown> & { module: string }
+  ): Logger => {
+    const child = this._logger.child(bindings)
+    return new Logger(child)
+  }
+
+  public error = (
+    // NOTE: Giving type as `unknown` here
+    // because when we do a `catch`,
+    // the caught error will be of type `unknown`
+    messageOrError: unknown
+  ): void => {
+    if (typeof messageOrError === "string") {
+      const stackFrames = get()
+      const stackTrace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
+
+      this._logger.error(
+        {
+          stackTrace,
+        },
+        messageOrError
+      )
+    } else if (messageOrError instanceof Error) {
+      const stackFrames = parse(messageOrError)
+      const stackTrace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
+
+      this._logger.error({
+        error: messageOrError,
+        stackTrace,
+      })
+    } else if (hasErrorProp(messageOrError)) {
+      const stackFrames = parse(messageOrError?.error)
+      const stackTrace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
+
+      this._logger.error({
+        error: messageOrError.error,
+        meta: _.omit(messageOrError, "error"),
+        stackTrace,
+      })
+    } else {
+      const stackFrames = get()
+      const stackTrace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
```

Comment:
Bit of repetition of the `slice()` calls, we could shift that into `getTraceMeta()`